[BUG] Expiremental stateful test runner gives false positives with check: ignored_auth

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

Same as: https://github.com/schemathesis/schemathesis/issues/2462 but for the expiremental test runner.
While using an openAPI schema with a securityscheme of type bearer, you receive `Authentication declared but not enforced for this operation` although authorization header is passed in by cli `--header` argument.


### To Reproduce

🚨 **Mandatory** 🚨: Steps to reproduce the behavior:

1. Create an open api definition file with bearer auth, on all endpoints.
```yaml
  securitySchemes:
    bearerAuth:
      type: http
      scheme: bearer
      bearerFormat: JWT
security:
  - bearerAuth: []
```
2. Run schemathesis in docker with flag:  `--experimental=stateful-test-runner --header "Authorization: Bearer {redactedToken}`
3.  The following test output is returned:
```
2024-09-26T06:34:58.801614601Z Stateful tests
2024-09-26T06:34:58.801616874Z 
2024-09-26T06:37:24.529131283Z F.....................................................................................................................................................
2024-09-26T06:37:32.835049844Z 
2024-09-26T06:37:32.835098198Z 
2024-09-26T06:37:32.835101059Z =================================== FAILURES ===================================
2024-09-26T06:37:32.835103995Z ________________________________ Stateful tests ________________________________
2024-09-26T06:37:32.836204839Z 1. Test Case ID: 698sSf
2024-09-26T06:37:32.836239843Z 
2024-09-26T06:37:32.836242895Z - Authentication declared but not enforced for this operation
2024-09-26T06:37:32.836245469Z 
2024-09-26T06:37:32.836247095Z     The API returned `200 OK` for `GET /redacted/redacted` with invalid auth.
2024-09-26T06:37:32.836249218Z 
2024-09-26T06:37:32.836251207Z [200] OK:
2024-09-26T06:37:32.836257250Z 
2024-09-26T06:37:32.836261364Z     `[{"id":"REDACTED // Output truncated...`
2024-09-26T06:37:32.836278103Z 
2024-09-26T06:37:32.836296030Z Reproduce with: 
2024-09-26T06:37:32.836310005Z 
2024-09-26T06:37:32.836317294Z     curl -X GET -H 'Authorization: [Filtered]' http://redacted:8080/redacted/redacted
2024-09-26T06:37:32.836323632Z 
2024-09-26T06:37:32.836325566Z =================================== SUMMARY ====================================
2024-09-26T06:37:32.836342159Z 
```
4. When not using the experimental stateful runner all checks pass. 

Please include a minimal API schema causing this issue:

```yaml
openapi: 3.1.0
security:
  - APIKey: []
  - UserAuth: []

paths:
  /dashboard:
    get:
      description: get trending clusters
      security:
        - UserAuth: []
      responses:
        '200':
          description: success
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object
                  properties:
                    size:
                      type: integer
                    name:
                      type: string
                    tags:
                      type: array
                      items:
                        type: object
        '401':
          $ref: "#/components/responses/401Unauthorized"
        '404':
          $ref: "#/components/responses/404NotFound"
        '500':
          description: Internal server error
 
components:
  securitySchemes:
    APIKey:
      type: http
      in: header
      scheme: bearer
      bearerFormat: APIKEY
    UserAuth:
      type: http
      in: header
      scheme: bearer
      bearerFormat: JWT

  responses:
    404NotFound:
      description: "Not found response"
    401Unauthorized:
      description: authorization error
      content:
        application/json:
          schema:
            type: object
            properties:
              error:
                type: string
    405MethodNotAllowed:
      description: method not allowed
```

### Expected behavior

When the correct authorization header is provided

### Environment

```
- OS: Windows 11, wsl2,  Docker version 26.1.4, build 5650f9b
- Python version: docker
- Schemathesis version: 3.36.3
- Spec version: 3.0.3 and 3.1.0
```

### Additional context

Include any other relevant details, like logs or screenshots.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -4,6 +4,11 @@ Changelog
 :version:`Unreleased <v3.36.3...HEAD>` - TBD
 --------------------------------------------
 
+**Fixed**
+
+- False positive for ``ignored_auth`` when used in stateful test runner. :issue:`2482`
+- Open Api 3.1 spec using ``$ref`` in a path is incorrectly validated as invalid. :issue:`2484`
+
 .. _v3.36.3:
 
 :version:`3.36.3 <v3.36.2...v3.36.3>` - 2024-09-29
@@ -14,9 +19,6 @@ Changelog
 - Meta information about generated data in the coverage phase.
 
 .. _v3.36.2:
-
-**Fixed**
-- OpenApi3.1 spec using $ref in a path is incorrectly validated as invalid
 
 :version:`3.36.2 <v3.36.1...v3.36.2>` - 2024-09-26
 --------------------------------------------------

--- a/src/schemathesis/internal/checks.py
+++ b/src/schemathesis/internal/checks.py
@@ -6,6 +6,7 @@ from dataclasses import dataclass
 from typing import TYPE_CHECKING, Callable, Optional
 
 if TYPE_CHECKING:
+    from requests.auth import HTTPDigestAuth
     from requests.structures import CaseInsensitiveDict
 
     from ..models import Case
@@ -23,7 +24,7 @@ class CheckContext:
     Provides access to broader test execution data beyond individual test cases.
     """
 
-    auth: RawAuth | None = None
+    auth: HTTPDigestAuth | RawAuth | None = None
     headers: CaseInsensitiveDict | None = None
 
 

--- a/src/schemathesis/stateful/runner.py
+++ b/src/schemathesis/stateful/runner.py
@@ -12,6 +12,7 @@ from hypothesis.control import current_build_context
 from hypothesis.errors import Flaky, Unsatisfiable
 
 from ..exceptions import CheckFailed
+from ..internal.checks import CheckContext
 from ..targets import TargetMetricCollector
 from . import events
 from .config import StatefulTestRunnerConfig
@@ -113,6 +114,7 @@ def _execute_state_machine_loop(
 ) -> None:
     """Execute the state machine testing loop."""
     from hypothesis import reporting
+    from requests.structures import CaseInsensitiveDict
 
     from ..transports import RequestsTransport
 
@@ -129,6 +131,7 @@ def _execute_state_machine_loop(
         if config.auth is not None:
             session.auth = config.auth
         call_kwargs["session"] = session
+    check_ctx = CheckContext(auth=config.auth, headers=CaseInsensitiveDict(config.headers) if config.headers else None)
 
     class _InstrumentedStateMachine(state_machine):  # type: ignore[valid-type,misc]
         """State machine with additional hooks for emitting events."""
@@ -223,7 +226,8 @@ def _execute_state_machine_loop(
             validate_response(
                 response=response,
                 case=case,
-                ctx=ctx,
+                runner_ctx=ctx,
+                check_ctx=check_ctx,
                 checks=config.checks,
                 additional_checks=additional_checks,
                 max_response_time=config.max_response_time,

--- a/src/schemathesis/stateful/validation.py
+++ b/src/schemathesis/stateful/validation.py
@@ -1,6 +1,6 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING
 
 from ..exceptions import CheckFailed, get_grouped_exception
 from ..internal.checks import CheckContext
@@ -17,27 +17,24 @@ def validate_response(
     *,
     response: GenericResponse,
     case: Case,
-    ctx: RunnerContext,
+    runner_ctx: RunnerContext,
+    check_ctx: CheckContext,
     checks: tuple[CheckFunction, ...],
     additional_checks: tuple[CheckFunction, ...] = (),
     max_response_time: int | None = None,
-    headers: dict[str, Any] | None = None,
 ) -> None:
     """Validate the response against the provided checks."""
-    from requests.structures import CaseInsensitiveDict
-
     from .._compat import MultipleFailures
     from ..checks import _make_max_response_time_failure_message
     from ..failures import ResponseTimeExceeded
     from ..models import Check, Status
 
     exceptions: list[CheckFailed | AssertionError] = []
-    check_results = ctx.checks_for_step
-    check_ctx = CheckContext(headers=CaseInsensitiveDict(headers) if headers else None)
+    check_results = runner_ctx.checks_for_step
 
     def _on_failure(exc: CheckFailed | AssertionError, message: str, context: FailureContext | None) -> None:
         exceptions.append(exc)
-        if ctx.is_seen_in_suite(exc):
+        if runner_ctx.is_seen_in_suite(exc):
             return
         failed_check = Check(
             name=name,
@@ -49,9 +46,9 @@ def validate_response(
             context=context,
             request=None,
         )
-        ctx.add_failed_check(failed_check)
+        runner_ctx.add_failed_check(failed_check)
         check_results.append(failed_check)
-        ctx.mark_as_seen_in_suite(exc)
+        runner_ctx.mark_as_seen_in_suite(exc)
 
     def _on_passed(_name: str, _case: Case) -> None:
         passed_check = Check(
@@ -72,16 +69,16 @@ def validate_response(
             if not skip_check:
                 _on_passed(name, copied_case)
         except CheckFailed as exc:
-            if ctx.is_seen_in_run(exc):
+            if runner_ctx.is_seen_in_run(exc):
                 continue
             _on_failure(exc, str(exc), exc.context)
         except AssertionError as exc:
-            if ctx.is_seen_in_run(exc):
+            if runner_ctx.is_seen_in_run(exc):
                 continue
             _on_failure(exc, str(exc) or f"Custom check failed: `{name}`", None)
         except MultipleFailures as exc:
             for subexc in exc.exceptions:
-                if ctx.is_seen_in_run(subexc):
+                if runner_ctx.is_seen_in_run(subexc):
                     continue
                 _on_failure(subexc, str(subexc), subexc.context)
 
@@ -93,7 +90,7 @@ def validate_response(
             try:
                 raise AssertionError(message)
             except AssertionError as _exc:
-                if not ctx.is_seen_in_run(_exc):
+                if not runner_ctx.is_seen_in_run(_exc):
                     _on_failure(_exc, message, context)
         else:
             _on_passed("max_response_time", case)

--- a/test/specs/openapi/stateful/conftest.py
+++ b/test/specs/openapi/stateful/conftest.py
@@ -5,7 +5,7 @@ from dataclasses import dataclass
 
 import hypothesis
 import pytest
-from flask import Flask, jsonify, request
+from flask import Flask, abort, jsonify, request
 
 import schemathesis
 from schemathesis.stateful.runner import StatefulTestRunnerConfig
@@ -22,6 +22,8 @@ class AppConfig:
     unsatisfiable: bool = False
     custom_headers: dict | None = None
     multiple_source_links: bool = False
+    auth_token: str | None = None
+    ignored_auth: bool = False
     slowdown: float | int | None = None
 
 
@@ -267,6 +269,26 @@ def app_factory(empty_open_api_3_schema):
             time.sleep(config.slowdown)
         return jsonify({"message": "Nothing happened"}), 200
 
+    @app.before_request
+    def check_auth():
+        if config.ignored_auth or config.auth_token is None or request.endpoint == get_spec.__name__:
+            # Allow all requests if auth is ignored or no token is set + to schema
+            return
+
+        auth_header = request.headers.get("Authorization")
+        if not auth_header:
+            abort(401, description="Authorization header is missing")
+
+        try:
+            token_type, token = auth_header.split()
+            if token_type.lower() != "bearer":
+                abort(401, description="Invalid token type")
+
+            if token != config.auth_token:
+                abort(401, description="Invalid token")
+        except ValueError:
+            abort(401, description="Invalid Authorization header format")
+
     def _factory(
         *,
         use_after_free=False,
@@ -279,10 +301,20 @@ def app_factory(empty_open_api_3_schema):
         custom_headers=None,
         multiple_source_links=False,
         single_link=False,
+        auth_token=None,
+        ignored_auth=False,
         slowdown=None,
     ):
         config.use_after_free = use_after_free
         config.ensure_resource_availability = ensure_resource_availability
+        config.auth_token = auth_token
+        config.ignored_auth = ignored_auth
+        if ignored_auth:
+            empty_open_api_3_schema["components"]["securitySchemes"] = {
+                "bearerAuth": {"type": "http", "scheme": "bearer", "bearerFormat": "JWT"}
+            }
+            empty_open_api_3_schema["security"] = [{"bearerAuth": []}]
+
         config.merge_body = merge_body
         if not merge_body:
             empty_open_api_3_schema["paths"]["/users"]["post"]["responses"]["201"]["links"]["UpdateUser"][

--- a/test/specs/openapi/stateful/test_runner.py
+++ b/test/specs/openapi/stateful/test_runner.py
@@ -13,7 +13,7 @@ from schemathesis.extra._flask import run_server
 from schemathesis.generation import DataGenerationMethod
 from schemathesis.internal.copy import fast_deepcopy
 from schemathesis.service.serialization import _serialize_stateful_event
-from schemathesis.specs.openapi.checks import response_schema_conformance, use_after_free
+from schemathesis.specs.openapi.checks import ignored_auth, response_schema_conformance, use_after_free
 from schemathesis.stateful.config import StatefulTestRunnerConfig
 from schemathesis.stateful.runner import events
 from schemathesis.stateful.sink import StateMachineSink
@@ -695,3 +695,34 @@ def test_unique_data(runner_factory):
             assert len(cases) == len(set(cases)), "Duplicate cases found"
         elif isinstance(event, events.RunFinished):
             assert event.status == events.RunStatus.FAILURE
+
+
+def test_ignored_auth_valid(runner_factory):
+    # When auth works properly
+    token = "Test"
+    runner = runner_factory(
+        app_kwargs={"auth_token": token},
+        config_kwargs={
+            "checks": (ignored_auth,),
+            "headers": {"Authorization": f"Bearer {token}"},
+        },
+    )
+    result = collect_result(runner)
+    # Then no failures are reported
+    assert result.events[-1].status == events.RunStatus.SUCCESS
+
+
+def test_ignored_auth_invalid(runner_factory):
+    # When auth is ignored
+    token = "Test"
+    runner = runner_factory(
+        app_kwargs={"auth_token": token, "ignored_auth": True},
+        config_kwargs={
+            "checks": (ignored_auth,),
+            "headers": {"Authorization": "Bearer UNKNOWN"},
+        },
+    )
+    result = collect_result(runner)
+    # Then it should be reported
+    assert result.events[-1].status == events.RunStatus.FAILURE
+    assert result.failures[0].message == "Authentication declared but not enforced for this operation"

--- a/test/specs/openapi/test_ignored_auth.py
+++ b/test/specs/openapi/test_ignored_auth.py
@@ -255,6 +255,82 @@ def test_explicit_auth_cli(cli, schema_url, snapshot_cli):
     assert cli.run(schema_url, "-c", "ignored_auth", "--auth=test:test", "--hypothesis-max-examples=1") == snapshot_cli
 
 
+@pytest.mark.openapi_version("3.0")
+@pytest.mark.parametrize("with_error", [True, False])
+@pytest.mark.snapshot(replace_statistic=True)
+def test_stateful_in_cli_no_error(testdir, cli, with_error, base_url, snapshot_cli):
+    target = "ignored" if with_error else "valid"
+    schema = {
+        "info": {"description": "An API to test Schemathesis", "title": "Example API", "version": "1.0.0"},
+        "openapi": "3.0.2",
+        "paths": {
+            "/basic": {
+                "get": {
+                    "operationId": "valid",
+                    "security": [{"basicAuth": []}],
+                    "responses": {"200": {"description": "OK"}},
+                }
+            },
+            "/ignored_auth": {
+                "get": {
+                    "operationId": "ignored",
+                    "security": [{"heisenAuth": []}],
+                    "responses": {"200": {"description": "OK"}},
+                },
+            },
+            "/users/": {
+                "post": {
+                    "requestBody": {
+                        "content": {
+                            "application/json": {
+                                "schema": {
+                                    "type": "object",
+                                    "properties": {
+                                        "first_name": {"type": "string", "minLength": 3},
+                                        "last_name": {"type": "string", "minLength": 3},
+                                    },
+                                    "required": ["first_name", "last_name"],
+                                    "additionalProperties": False,
+                                },
+                            }
+                        },
+                        "required": True,
+                    },
+                    "responses": {
+                        "201": {
+                            "description": "OK",
+                            "links": {
+                                "TestLink": {"operationId": target, "parameters": {}},
+                            },
+                        }
+                    },
+                }
+            },
+        },
+        "components": {
+            "securitySchemes": {
+                "basicAuth": {"scheme": "basic", "type": "http"},
+                "heisenAuth": {"scheme": "basic", "type": "http"},
+            },
+        },
+    }
+    schema_file = testdir.make_openapi_schema_file(schema)
+    assert (
+        cli.run(
+            str(schema_file),
+            f"--base-url={base_url}",
+            "-c",
+            "ignored_auth",
+            "--header=Authorization: Basic dGVzdDp0ZXN0",
+            "--hypothesis-max-examples=100",
+            "--experimental=stateful-only",
+            "--experimental=stateful-test-runner",
+            "--show-trace",
+        )
+        == snapshot_cli
+    )
+
+
 @pytest.mark.skipif(sys.version_info < (3, 10), reason="Typing syntax is not supported on Python 3.9 and below")
 def test_custom_auth():
     app = FastAPI()

